### PR TITLE
Added voting endpoint

### PR DESCRIPTION
**Pull request by vctfernandes**
Hey guys,

Just implemented voting endpoint for my needs.
Looks good enough, but I'm open to suggestions, of course 😄

### CODE REVIEW

**Comment by vctfernandes** [comment]
@julianlam, do you believe this needs improvements before it can be merged?

**Comment by julianlam** [comment]
Hey there, looks like this feel out of my inbox, will take a look 😄

**Comment by julianlam** [comment]
I think I would prefer:

* `POST` upvote or downvote (passing in `delta`, I suppose? And then you just check whether it's > 0 or < 0)
* `DELETE` unvte

Subsequent `POST`s should do nothing (unlike the UI which unvotes).

... also please duplicate these routes into v2 😄

**Comment by vctfernandes** [comment]
Haven't had much free time these past weeks, sorry!
Let me see if I understand correctly what you're proposing:
POST /:pid/upvote (if delta < 0, treat as downvote)
POST /:pid/downvote (if delta < 0, treat as upvote)
DELETE /:pid/unvote 

Something like this?

**Comment by julianlam** [comment]
@vctfernandes you'd want to use the route `/api/v1/topics/<tid>/vote`

**Comment by vctfernandes** [comment]
Doesn't `/api/v1/topics/<pid>/vote` makes more sense? Given that votes are post related and not topic related.

**Comment by julianlam** [comment]
Yes my apology, you are correct

**Comment by vctfernandes** [comment]
@julianlam have a look 😃

File: routes/v1/posts.js [hunk 1]
```diff
@@ -38,7 +38,46 @@ module.exports = function(middleware) {
...
+			if (!utils.checkRequired(['delta'], req, res)) {
+				return false;
+			}
+			if (typeof req.body.delta !== 'number') {
+				res.status(400).json(errorHandler.generate(
+					400, 'invalid-params',
+					'Required parameters were used incorrectly in this API call, please see the "params" property',
+					['delta']
+				));
+				return false;
```
Comment (julianlam) [inline routes/v1/posts.js]: I think you can use `utils.checkRequired` here:

```
if (!utils.checkRequired(['delta'], req, res)) {
	return false;
}
```

File: routes/v1/posts.js [hunk 2]
```diff
@@ -38,7 +38,46 @@ module.exports = function(middleware) {
...
+			}
+
+			var payload = {
+				uid: req.user.uid,
+				pid: req.params.pid,
+				delta: req.body.delta
+			}
```
Comment (julianlam) [inline routes/v1/posts.js]: No specific need to define `payload` if it is not used in its entirety, looks like later on you just use the properties on their own.

File: routes/v2/posts.js [hunk 3]
```diff
@@ -49,8 +49,47 @@ module.exports = function(middleware) {
...
+				return false;
+			}
+			if (typeof req.body.delta !== 'number') {
+				res.status(400).json(errorHandler.generate(
+					400, 'invalid-params',
+					'Required parameters were used incorrectly in this API call, please see the "params" property',
+					['delta']
+				));
+				return false;
+			}
```
Comment (julianlam) [inline routes/v2/posts.js]: Same across v2 please :smile:

File: routes/v1/readme.md [hunk 4]
```diff
@@ -120,6 +120,13 @@ an administrative uid. All other uids passed in will result in an error.
         * `DELETE /:pid`
             * Deletes a post (**Careful**: There is no confirmation!)
             * **Accepts**: No parameters
+        * `POST /:pid/vote`
+            * Votes for a post
+            * **Requires**: `delta`
+            * `delta` must be a number. If `delta > 0`, it's considered an upvote; otherwise, it's a downvote.
```
Comment (julianlam) [inline routes/v1/readme.md]: Do you think there are libraries that cannot send `DELETE`s, and only `POST`s? Maybe if it is 0, you can count it as an unvote :stuck_out_tongue:

**Comment by vctfernandes** [comment]
Made the requested changes, hope it's fine now 😁

**Comment by julianlam** [comment]
LGTM! v5.0.2

**Pull request merged**